### PR TITLE
Fix FuncOp lowering.

### LINKED ISSUE #57: Assertion error with trivial kernel
```
import triton


@triton.jit
def kernel(x):
    pass

kernel[(1, )](2)
```

Running this triggers an "index out of bounds error". It seems to boil down to [LLVM expecting that `funcOp.getAllArgAttrs().size() == funcOp.getNumArguments()`](https://github.com/llvm/llvm-project/blob/a778909168746e266ad52b817a758328cdd28311/mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp#L341-L346), which seems to make sense, except that our `kernel`'s FuncOp somehow has `getNumArguments() == 1` but `getAllArgAttrs() == nullptr` (which corresponds to no arg attrs at all). Not sure what's responsible for populating the arg attrs...

### PR DESCRIPTION
**Pull request by ienkovich**
Resolves #57

### CODE REVIEW

**Review by minjang — APPROVED** [review]
Thanks for the quick fix!

File: third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp [hunk 1]
```diff
@@ -88,16 +88,18 @@ struct FuncOpConversion : public ConvertOpToLLVMPattern<triton::FuncOp> {
     SmallVector<NamedAttribute> amendedAttrs;
     filterFuncAttributes(funcOp, /*filterArgAttrs=*/true, amendedAttrs);
     SmallVector<Attribute> amendedArgAttrs;
-    if (funcOp.getAllArgAttrs())
+    if (funcOp.getAllArgAttrs()) {
```
Comment (int3) [inline third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp:91]: I wonder why this check isn't necessary in the corresponding GPU pass:

https://github.com/triton-lang/triton-cpu/blob/22544ed8dd963726e738e571108c7fc05114da8f/lib/Conversion/TritonGPUToLLVM/FuncOpToLLVM.cpp#L56-L59

File: third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp [hunk 2]
```diff
@@ -88,16 +88,18 @@ struct FuncOpConversion : public ConvertOpToLLVMPattern<triton::FuncOp> {
...
+      amendedArgAttrs.emplace_back(DictionaryAttr::get(ctx));
+      amendedArgAttrs.emplace_back(DictionaryAttr::get(ctx));
+      amendedArgAttrs.emplace_back(DictionaryAttr::get(ctx));
+      amendedArgAttrs.emplace_back(DictionaryAttr::get(ctx));
+      amendedArgAttrs.emplace_back(DictionaryAttr::get(ctx));
+      amendedAttrs.push_back(
+          rewriter.getNamedAttr(funcOp.getArgAttrsAttrName(),
+                                rewriter.getArrayAttr(amendedArgAttrs)));
+    }
     // 3. Add a new arguments to the region
```
Comment (int3) [inline third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp:103]: if we aren't adding the extra args above, I guess we should skip the calls to `region.addArgument` below?

but also... is it not possible that a kernel that has no arg attrs still wants to access the program IDs?

(also: why are program IDs passed as arguments for CPU kernels but not for GPU ones?)

Comment (minjang) [reply third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp:103]: > why are program IDs passed as arguments for CPU kernels but not for GPU ones?

For GPU, program ids are directly retrieved via `%ctaid` registers. For CPU, pid is determined by OpenMP's parallel-for loop iteration index. So, we pass pids as the additional parameters.

Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp:103]: We always add extra args, but arg attrs are added conditionally now. GPU backends have a similar code to add extra args for program ids and shared memory pointer.

File: third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp (hunk 1 above)
Comment (ienkovich) [reply third_party/cpu/lib/TritonCPUToLLVM/FuncOpToLLVM.cpp:91]: I guess GPU backend might have the same issue but it is simply not triggered by any of the current tests.

**Pull request merged**